src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    auto const optSubj = tx[~sfSubject];
+    auto const optIss = tx[~sfIssuer];
+    auto const optSig = tx[~sfSignature];
+
+    if ((optSubj.has_value() == optIss.has_value()) ||
+        (optIss.has_value() != optSig.has_value()))
+    {
+        // Either both fields are present or neither field is present.  In
+        // either case the transaction is malformed.
+        JLOG(j.trace())
+            << "Malformed transaction: "
+               "Invalid Subject, Issuer and Signature fields combination.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    AccountID const target{optSubj ? *optSubj : *optIss};
+    if (!target)
+    {
+        JLOG(j.trace()) << "Malformed transaction: Subject or Issuer "
+                           "field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    auto const optUri = tx[~sfURI];
+    if (optUri && (optUri->size() > maxCredentialURILength))
+    {
+        JLOG(j.trace()) << "Malformed transaction: "
+                           "URI too long.";
+        return temMALFORMED;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialCreate::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const acc{ctx.tx[sfAccount]};
+
+    auto const& txCredType(ctx.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    if (ctx.tx.isFieldPresent(sfSubject))
+    {
+        AccountID const subj{ctx.tx[sfSubject]};
+        if (!ctx.view.exists(keylet::account(subj)))
+            return tecNO_TARGET;
+        if (ctx.view.exists(keylet::credential(subj, acc, credType)))
+            return tecDUPLICATE;
+    }
+    else if (ctx.tx.isFieldPresent(sfIssuer))
+    {
+        AccountID const issuer{ctx.tx[sfIssuer]};
+        if (!ctx.view.exists(keylet::account(issuer)))
+            return tecNO_TARGET;
+        if (ctx.view.exists(keylet::credential(acc, issuer, credType)))
+            return tecDUPLICATE;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialCreate::doApply()
+{
+    std::uint32_t const closeTime =
+        ctx_.view().info().parentCloseTime.time_since_epoch().count();
+
+    auto const sleOwner = view().peek(keylet::account(account_));
+    if (!sleOwner)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        STAmount const reserve{view().fees().accountReserve(
+            sleOwner->getFieldU32(sfOwnerCount) + 1)};
+
+        if (mPriorBalance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    auto const& txCredType(ctx_.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    AccountID const subj = ctx_.tx.isFieldPresent(sfSubject)
+        ? AccountID(ctx_.tx[sfSubject])
+        : account_;
+    bool const issPresent = ctx_.tx.isFieldPresent(sfIssuer);
```

Comment:
This boolean is only used for one ternary - I'd prefer if a separate variable wasn't created just for that.

Reply:
removed